### ISSUE TITLE
Pyright does not seem to correctly narrow unions of type objects

### ISSUE THREAD

**Issue by tylerhou** [microsoft/pyright#7503]
Pyright does not seem to narrow union types when the members of the union type are themselves `type[...]`.

Code sample in [pyright playground 1.1.354](https://pyright-play.net/?pyrightVersion=1.1.354&code=JYWwDg9gTgLgBFAhgOwCYRAKAGZQ3GATzGGQHM5RJY5EBnOgU1gH1lGA3ZzTAYwBt6dOAEEAXJjhS4YITwFC4AIQnSZcvgAsIwXozgBeBCnQgAdL227GACgDaIgDTKAugEpMwbHEs69ho3FJaTAoUhgbACIRSI9Gfi8fK38DIxVgqVDwqKVYzHimVWkhZhg2TmYbX2s3IA)

```python
import random
from typing import assert_never

class A:
    pass

class B:
    pass

choice = random.choice([A, B])
if choice == A:
    print("A")
elif choice == B:
    print("B")
else:
    assert_never(choice)
```

On the `assert_never` line, Pyright reports the following error:

```
Argument of type "type[A] | type[B]" cannot be assigned to parameter "__arg" of type "Never" in function "assert_never"
  Type "type[A] | type[B]" cannot be assigned to type "Never"
    Type "type[A]" cannot be assigned to type "Never"  (reportArgumentType)
```

My assumption is that the correct behavior should be no error because `type[A] | type[B]` should have been narrowed to `Never` by the branches.

**Comment by erictraut** [comment]
This is not a supported type guard pattern. For a full list of supported patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards). 

We do occasionally add support for new patterns, but each one requires additional custom logic and can impact analysis speed overall, so we do so only when there's strong signal from pyright users that it's a common pattern. This is the first time that I've heard of this pattern being used. Plus, it's not supported by other type checkers like mypy. For that reason, I don't think this is common enough to justify supporting at this time. I'm therefore going to close this as "not planned". I'm willing to revisit this decision in the future if there's additional signal (e.g. in the form of upvotes).

**Comment by erictraut** [comment]
Here is a workaround that narrows as expected:

```python
if issubclass(choice, A):
    print("A")
elif issubclass(choice, B):
    print("B")
else:
    assert_never(choice)
```

**Comment by tylerhou** [comment]
Thanks for the response; makes sense. I will use the workaround.

The context where I am using this is for fuzzing: to generate random ASTs, I randomly choose to create e.g. a binary expression or an integer literal. My AST nodes are classes and I am using `random.choice` to pick one to instantiate.
